ConcurrentModificationException on the second commit updating a multi-valued object property

I'm using Jopa 2.1.0 with ontodriver-jena.  My model consists of a `Child` class, and a `Parent` class with a property `hasChild` with a `Set` of `Child` as range. Once a `Parent` instance `p` is stored in the ontology, I need to add several instances of `Child` as children of `p`, one by one and using different transactions. Using an in-memory storage, two additions works well. Instead, using a file storage, only the first addition works fine, where as the second one cause the runtime exception reported below.

A test showing this use case both with in-memory and file storage is available at https://github.com/cristianolongoodhl/jopamiltivaluedpropertiesissue.

Also, workaround suggestions are welcome.

`cz.cvut.kbss.jopa.exceptions.RollbackException: java.util.ConcurrentModificationException

	at cz.cvut.kbss.jopa.transactions.EntityTransactionWrapper.commit(EntityTransactionWrapper.java:56)
	at cz.cvut.kbss.jopa.transactions.EntityTransactionImpl.commit(EntityTransactionImpl.java:58)
	at it.unict.gallosiciliani.importing.persistence.LexiconOntologyWriterTest.testStoringMultipleEntriesAtomic(LexiconOntologyWriterTest.java:169)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.util.ConcurrentModificationException
	at org.apache.jena.mem.ArrayBunch$2.forEachRemaining(ArrayBunch.java:130)
	at org.apache.jena.util.iterator.WrappedIterator.forEachRemaining(WrappedIterator.java:113)
	at org.apache.jena.util.iterator.FilterIterator.forEachRemaining(FilterIterator.java:97)
	at org.apache.jena.util.iterator.WrappedIterator.forEachRemaining(WrappedIterator.java:113)
	at org.apache.jena.mem.TrackingTripleIterator.forEachRemaining(TrackingTripleIterator.java:58)
	at org.apache.jena.util.iterator.Map1Iterator.forEachRemaining(Map1Iterator.java:54)
	at org.apache.jena.util.iterator.WrappedIterator.forEachRemaining(WrappedIterator.java:113)
	at cz.cvut.kbss.ontodriver.jena.connector.LocalStorage.remove(LocalStorage.java:123)
	at cz.cvut.kbss.ontodriver.jena.connector.SharedStorageConnector.lambda$removePropertyValues$6(SharedStorageConnector.java:160)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at cz.cvut.kbss.ontodriver.jena.connector.SharedStorageConnector.removePropertyValues(SharedStorageConnector.java:158)
	at cz.cvut.kbss.ontodriver.jena.connector.ChangeTrackingStorageConnector.mergeRemovedStatements(ChangeTrackingStorageConnector.java:80)
	at cz.cvut.kbss.ontodriver.jena.connector.ChangeTrackingStorageConnector.commit(ChangeTrackingStorageConnector.java:59)
	at cz.cvut.kbss.ontodriver.jena.JenaAdapter.commit(JenaAdapter.java:66)
	at cz.cvut.kbss.ontodriver.jena.JenaConnection.commit(JenaConnection.java:66)
	at cz.cvut.kbss.jopa.sessions.ConnectionWrapper.commit(ConnectionWrapper.java:103)
	at cz.cvut.kbss.jopa.sessions.ChangeTrackingUnitOfWork.commitToStorage(ChangeTrackingUnitOfWork.java:114)
	at cz.cvut.kbss.jopa.sessions.AbstractUnitOfWork.commitUnitOfWork(AbstractUnitOfWork.java:199)
	at cz.cvut.kbss.jopa.sessions.AbstractUnitOfWork.commit(AbstractUnitOfWork.java:190)
	at cz.cvut.kbss.jopa.transactions.EntityTransactionWrapper.commit(EntityTransactionWrapper.java:53)
`

--- a/ontodriver-jena/src/main/java/cz/cvut/kbss/ontodriver/jena/connector/FileStorage.java
+++ b/ontodriver-jena/src/main/java/cz/cvut/kbss/ontodriver/jena/connector/FileStorage.java
@@ -22,6 +22,8 @@ import cz.cvut.kbss.ontodriver.exception.OntoDriverInitializationException;
 import cz.cvut.kbss.ontodriver.jena.exception.JenaDriverException;
 import org.apache.jena.query.DatasetFactory;
 import org.apache.jena.rdf.model.Model;
+import org.apache.jena.rdf.model.Statement;
+import org.apache.jena.rdf.model.StmtIterator;
 import org.apache.jena.riot.RDFDataMgr;
 import org.apache.jena.riot.RDFLanguages;
 import org.apache.jena.riot.RiotNotFoundException;
@@ -31,6 +33,7 @@ import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.util.List;
 
 /**
  * File storage accessor.
@@ -97,5 +100,13 @@ class FileStorage extends LocalStorage {
         }
         dataset.close();
         initialize();
+    }
+
+    @Override
+    public void remove(StmtIterator iterator, String context) {
+        // For FileStorage we can't iterate through the statement and call remove on them because
+        // it causes a ConcurrentModificationException
+        final List<Statement> toRemove = iterator.toList();
+        remove(toRemove, context);
     }
 }

--- a/ontodriver-jena/src/test/java/cz/cvut/kbss/ontodriver/jena/connector/StorageTest.java
+++ b/ontodriver-jena/src/test/java/cz/cvut/kbss/ontodriver/jena/connector/StorageTest.java
@@ -23,16 +23,23 @@ import cz.cvut.kbss.ontodriver.jena.config.JenaConfigParam;
 import cz.cvut.kbss.ontodriver.jena.config.JenaOntoDriverProperties;
 import cz.cvut.kbss.ontodriver.jena.environment.Generator;
 import org.apache.jena.query.DatasetFactory;
+import org.apache.jena.query.ReadWrite;
+import org.apache.jena.rdf.model.RDFNode;
 import org.apache.jena.rdf.model.ResourceFactory;
+import org.apache.jena.rdf.model.StmtIterator;
 import org.apache.jena.vocabulary.RDF;
 import org.junit.jupiter.api.Test;
 
 import java.io.File;
 import java.nio.file.Files;
 
+import static org.apache.jena.rdf.model.ResourceFactory.createResource;
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 
 public class StorageTest extends StorageTestUtil {
 
@@ -42,7 +49,7 @@ public class StorageTest extends StorageTestUtil {
         config.setProperty(JenaConfigParam.STORAGE_TYPE, JenaOntoDriverProperties.IN_MEMORY);
         final Storage result = Storage.create(config);
         assertNotNull(result);
-        assertTrue(result instanceof MemoryStorage);
+        assertInstanceOf(MemoryStorage.class, result);
         assertNotNull(result.getDataset());
     }
 
@@ -54,7 +61,7 @@ public class StorageTest extends StorageTestUtil {
         config.setProperty(JenaConfigParam.STORAGE_TYPE, JenaOntoDriverProperties.FILE);
         final Storage result = Storage.create(config);
         assertNotNull(result);
-        assertTrue(result instanceof FileStorage);
+        assertInstanceOf(FileStorage.class, result);
     }
 
     @Test
@@ -71,7 +78,7 @@ public class StorageTest extends StorageTestUtil {
         final DriverConfiguration config = createConfiguration("test:uri");
         final Storage result = Storage.create(config);
         assertNotNull(result);
-        assertTrue(result instanceof MemoryStorage);
+        assertInstanceOf(MemoryStorage.class, result);
     }
 
     @Test
@@ -96,5 +103,64 @@ public class StorageTest extends StorageTestUtil {
         final Storage storage = Storage.create(config);
         assertNotNull(storage);
         assertThrows(UnsupportedOperationException.class, () -> storage.setDataset(DatasetFactory.create()));
+    }
+
+    // Bug #274
+    @Test
+    void removeHandlesStatementIteratorForInMemoryStorage() {
+        final DriverConfiguration config = createConfiguration("test:uri");
+        config.setProperty(JenaConfigParam.STORAGE_TYPE, JenaOntoDriverProperties.IN_MEMORY);
+        final Storage storage = Storage.create(config);
+        storage.begin(ReadWrite.WRITE);
+        generateTestData(storage.getDataset());
+        storage.commit();
+
+        storage.begin(ReadWrite.WRITE);
+        final StmtIterator iterator = storage.getDefaultGraph()
+                                             .listStatements(createResource(TYPE_ONE), null, (RDFNode) null);
+        storage.remove(iterator, null);
+        storage.commit();
+    }
+
+    // Bug #274
+    @Test
+    void removeHandlesStatementIteratorForLocalStorage() throws Exception {
+        final File file = Files.createTempFile("jena-onto", ".ttl").toFile();
+        file.deleteOnExit();
+        final DriverConfiguration config = createConfiguration(file.getAbsolutePath());
+        config.setProperty(JenaConfigParam.STORAGE_TYPE, JenaOntoDriverProperties.FILE);
+        final Storage storage = Storage.create(config);
+        storage.begin(ReadWrite.WRITE);
+        generateTestData(storage.getDataset());
+        storage.commit();
+
+        storage.begin(ReadWrite.WRITE);
+        final StmtIterator iterator = storage.getDefaultGraph()
+                                             .listStatements(createResource(TYPE_ONE), null, (RDFNode) null);
+        storage.remove(iterator, null);
+        storage.commit();
+    }
+
+    // Bug #274
+    @Test
+    void removeHandlesStatementIteratorForTDB2Storage() throws Exception {
+        final File file = Files.createTempDirectory("jena-onto").toFile();
+        file.deleteOnExit();
+        try {
+            final DriverConfiguration config = createConfiguration(file.getAbsolutePath());
+            config.setProperty(JenaConfigParam.STORAGE_TYPE, JenaOntoDriverProperties.TDB2);
+            final Storage storage = Storage.create(config);
+            storage.begin(ReadWrite.WRITE);
+            generateTestData(storage.getDataset());
+            storage.commit();
+
+            storage.begin(ReadWrite.WRITE);
+            final StmtIterator iterator = storage.getDefaultGraph()
+                                                 .listStatements(createResource(TYPE_ONE), null, (RDFNode) null);
+            storage.remove(iterator, null);
+            storage.commit();
+        } finally {
+            deleteStorageDir(file);
+        }
     }
 }

